### PR TITLE
expression: implement vectorized evaluation for builtinMakeDateSig

### PR DESCRIPTION
**Pull request by ekalinin**
PCP #12101

### What problem does this PR solve? 

This PR implements vectorized `builtinMakeDateSig`

### What is changed and how it works?

```bash
➜ make vectorized-bench VB_FILE=Time VB_FUNC=builtinMakeDateSig
cd ./expression && \
	go test -v -benchmem \
		-bench=BenchmarkVectorizedBuiltinTimeFunc \
		-run=BenchmarkVectorizedBuiltinTimeFunc \
		-args "builtinMakeDateSig"
goos: linux
goarch: amd64
pkg: github.com/pingcap/tidb/expression
BenchmarkVectorizedBuiltinTimeFuncGenerated-12    	1000000000	         0.00908 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedBuiltinTimeFunc/builtinMakeDateSig-VecBuiltinFunc-12         	   12246	     94084 ns/op	       0 B/op	       0 allocs/op
BenchmarkVectorizedBuiltinTimeFunc/builtinMakeDateSig-NonVecBuiltinFunc-12      	    9542	    125687 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/pingcap/tidb/expression	3.337s
```

### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

File: expression/builtin_time_vec.go [hunk 1]
```diff
@@ -928,11 +928,71 @@ func (b *builtinAddDateIntDecimalSig) vecEvalTime(input *chunk.Chunk, result *ch
...
+		return err
+	}
+	defer b.bufAllocator.put(buf2)
+	if err := b.args[0].VecEvalInt(b.ctx, input, buf2); err != nil {
+		return err
+	}
+
+	result.ResizeTime(n, false)
+	result.MergeNulls(buf1)
+	result.MergeNulls(buf2)
```
Comment (SunRunAway) [inline expression/builtin_time_vec.go]: result.MergeNulls(buf1, buf2)

File: expression/builtin_time_vec_test.go [hunk 2]
```diff
@@ -87,7 +87,9 @@ var vecBuiltinTimeCases = map[string][]vecExprBenchCase{
 	ast.CurrentDate: {
 		{retEvalType: types.ETDatetime},
 	},
-	ast.MakeDate: {},
+	ast.MakeDate: {
+		{retEvalType: types.ETDatetime, childrenTypes: []types.EvalType{types.ETInt, types.ETInt}},
```
Comment (SunRunAway) [inline expression/builtin_time_vec_test.go]: I'd be better use `rangeInt64Gener`.

File: expression/builtin_time_vec.go [hunk 3]
```diff
@@ -928,11 +928,71 @@ func (b *builtinAddDateIntDecimalSig) vecEvalTime(input *chunk.Chunk, result *ch
...
+	if err := b.args[0].VecEvalInt(b.ctx, input, buf1); err != nil {
+		return err
+	}
+
+	buf2, err := b.bufAllocator.get(types.ETInt, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf2)
+	if err := b.args[0].VecEvalInt(b.ctx, input, buf2); err != nil {
```
Comment (wshwsh12) [inline expression/builtin_time_vec.go]:
```suggestion
	if err := b.args[1].VecEvalInt(b.ctx, input, buf2); err != nil {
```

Comment (ekalinin) [reply expression/builtin_time_vec.go]: Fixed.

File: expression/builtin_time_vec_test.go (hunk 2 above)
Comment (ekalinin) [reply expression/builtin_time_vec_test.go]: Done.

File: expression/builtin_time_vec.go (hunk 1 above)
Comment (ekalinin) [reply expression/builtin_time_vec.go]: Fixed.

File: expression/builtin_time_vec_test.go [hunk 4]
```diff
@@ -87,7 +87,11 @@ var vecBuiltinTimeCases = map[string][]vecExprBenchCase{
 	ast.CurrentDate: {
 		{retEvalType: types.ETDatetime},
 	},
-	ast.MakeDate: {},
+	ast.MakeDate: {
+		{retEvalType: types.ETDatetime, childrenTypes: []types.EvalType{types.ETInt, types.ETInt},
+			geners: []dataGenerator{&rangeInt64Gener{1700, 2200}, &rangeInt64Gener{0, 365}},
```
Comment (SunRunAway) [inline expression/builtin_time_vec_test.go]: I'm confused with the low bound of 1700,  IMO, the first argument can be smaller than 1700.

Comment (ekalinin) [reply expression/builtin_time_vec_test.go]: Fixed.

**Pull request merged**